Refactoring patches for endpoint and policystore

dianarg opened the pull request:
A few changes to clean up code related to geopmcd implementation and testing.

- Infinite timeout for Endpoint agent attach wait
- Changes to PolicyStore API to make agent consistently the first parameter
- Rename test_static_policy to test_enforce_policy
- Add Endpoint::wait_for_agent_detach()

dianarg committed:
Rename test_static_policy to test_enforce_policy

Change-Id: I09da04eb19d81056c31aedfb8b832073c1a99b13
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Add infinite timeout option for Endpoint agent wait

- Negative value for the timeout indicates endpoint will block forever
  until agent attaches or wait is cancelled.
- Resolves #983

Change-Id: Ic1c6036ff0267f2a2b8f2ea4cd564ca1f45dd5d0
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Update PolicyStore APIs to have agent as first parameter

Change-Id: I45bb291758e3fee8c48bb917cdb71850ea6b8503
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Add Endpoint wait_for_agent_detach()

- Intention is for daemon to use to avoid sending additional policies
  to the agent.

Change-Id: I9d865159764f80ff683ef04b76391b9428f565f5
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

cmcantalupo commented on src/Endpoint.cpp:
```diff
@@ -194,14 +194,26 @@ namespace geopm
+        std::string agent = get_agent();
+        double sample_age = 0;
+        std::vector<double> sample(m_num_sample);
+        while (m_continue_loop && agent != "" && sample_age < timeout) {
```
here timeout == 0 -> do not wait, in wait_for_agent_attach timeout == 0 -> wait forever.  This seems a little confusing.

dianarg replied:
The intention was that timeout==0 means do not wait.  I'll add a test for that.

dianarg replied:
So the condition needs to check the value of agent in wait_for_agent_attach

cmcantalupo commented on test/EndpointTest.cpp:
```diff
@@ -304,3 +304,27 @@ TEST_F(EndpointTest, wait_stops_when_agent_attaches)
     EXPECT_NE(result, std::future_status::timeout);
     mio.close();
 }
+
```
Does it make sense to test timeout==0 for the two wait methods?

dianarg committed:
Add tests and fix behavior of endpoint timeouts

- Timeout of zero means check condition and throw immediately without
  waiting.

Change-Id: I731f506910495f26035022fd10d18273e79e8260
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

cmcantalupo commented on src/Endpoint.cpp:
```diff
@@ -205,12 +205,16 @@ namespace geopm
             agent = get_agent();
-            sample_age = read_sample(sample);
             std::this_thread::sleep_for(std::chrono::milliseconds(100));
+            if (timeout >= 0 && geopm_time_since(&start) >= timeout) {
```
Not sure I understand what the check:

```timeout >= 0```

is doing.  Seems redundant.

cmcantalupo reviewed (commented):
Little nit, but I think this makes sense.
